niap-cc/Permissions/Tester/app/src/main/java/com/android/certifications/niap/permissions/SignaturePermissionTester.java

```diff
@@ -3656,18 +3875,625 @@ public void writeToParcel(Parcel parcel, int i) {
+                        throw new UnexpectedPermissionTestFailureException(e);
+                    }
+                }));
+
+        // # Skip GET_HISTORICAL_APP_OPS_STATS
+        // Reason : The corresponding api hasn't raise any exception internally.(AppOpsService#getHistoricalOps)
+
+        mPermissionTasks.put(permission.SET_SYSTEM_AUDIO_CAPTION,
+                new PermissionTest(false, Build.VERSION_CODES.TIRAMISU, () -> {
+                    invokeReflectionCall(mCaptioningManager.getClass(),
+                            "setSystemAudioCaptioningEnabled",mCaptioningManager,
+                            new Class<?>[]{boolean.class}, true);
+                }));
+
+
+        // # Skip INSTALL_DPC_PACKAGES
+        // Reason : Found no way to evaluate this permission without enterprise environment
+
+        //The Permission required for CTS test - Notification test suite - need platform signature to run
+
+//        mPermissionTasks.put(permission.REVOKE_POST_NOTIFICATIONS_WITHOUT_KILL,
+//                new PermissionTest(false, Build.VERSION_CODES.TIRAMISU, () -> {
+//                    @SuppressLint("WrongConstant") Object permissionManager = mContext.getSystemService("permission");
+//                    //revokePostNotificationPermissionWithoutKillForTest( java.lang.String int);
+//                     invokeReflectionCall(permissionManager.getClass(),
+//                            "revokePostNotificationPermissionWithoutKillForTest",permissionManager,
+//                            new Class<?>[]{String.class,int.class}, mContext.getPackageName(),0);
+//                }));
+
+
+        // # Skip DELIVER_COMPANION_MESSAGES
+        //Reason : This feature appears to have been punted to U. See b/199427116,
+
+        // # Skip MODIFY_TOUCH_MODE_STATE
+        //Reason : setInTouchMode has not raise any exception. no way to evaluate this permissison.
+
+        mPermissionTasks.put(permission.MODIFY_USER_PREFERRED_DISPLAY_MODE,
+                new PermissionTest(false, Build.VERSION_CODES.TIRAMISU, () -> {
+                    Display.Mode mode = mDisplayManager.getDisplays()[0].getMode();
+                    //The method may not find if the app is signing by the platform signing key
+                    //mLogger.logInfo(ReflectionUtils.checkDeclaredMethod(mDisplayManager,"setGlobalUser").toString());
+                    //invokeReflectionCall(mDisplayManager.getClass(),
+                    //        "setGlobalUserPreferredDisplayMode", mDisplayManager,
+                    //        new Class[]{Display.Mode.class},mode);
+
+                    mTransacts.invokeTransact(Transacts.DISPLAY_SERVICE,Transacts.DISPLAY_DESCRIPTOR,
+                    Transacts.setUserPreferredDisplayMode,0,mode);
+                }));
+
+        mPermissionTasks.put(permission.ACCESS_ULTRASOUND,
+                new PermissionTest(false, Build.VERSION_CODES.TIRAMISU, () -> {
+                    invokeReflectionCall(mAudioManager.getClass(),
```

Comment:
I noticed several reflection calls here; did you need to do anything to support these calls when the app is not signed by the platform key? I thought new APIs denied reflection calls by default.

Reply:
Some apis related to certain permissions are hidden to public (those are intended to use with system-app or system), 
But we need to use them to test some permissions within the scope of application.

With my understanding, some non-sdk interfaces are blocked, but it is organized.
We can still call those hidden but non-blocking apis.

https://developer.android.com/guide/app-compatibility/restrictions-non-sdk-interfaces

Reply:
Good to know, I thought all new hidden APIs were denied by default, glad some are still an option for this tester. Thanks!